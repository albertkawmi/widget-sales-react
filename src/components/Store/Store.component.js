import React from 'react'
import PropTypes from 'prop-types'
import api from '../../utils/api'

export default class Store extends React.Component {
  state = {
    clients: [],
    sales: [],
    error: '',
  }

  getChildContext = () => {
    const childContext = {
      ...this.state,
      fetchClients: () => this.request('clients'),
      fetchSales: () => this.request('sales'),
      fetchFault: () => this.request('fault'),
      dismissError: () => this.setState({ error: '' })
    }
    return childContext
  }

  request = endpoint => {
    return api.get(endpoint)
      .then(results => this.setState({
        [endpoint]: results
      }))
      .catch(error => this.setState({
        error: error.toString(),
      }))
  }

  render() {
    return this.props.children
  }
}

Store.childContextTypes = {
  clients: PropTypes.array,
  sales: PropTypes.array,
  error: PropTypes.string,
  fetchClients: PropTypes.func,
  fetchSales: PropTypes.func,
  fetchFault: PropTypes.func,
  dismissError: PropTypes.func
}
