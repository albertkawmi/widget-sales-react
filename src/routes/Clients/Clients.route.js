import React from 'react'
import ClientsTable from '../../components/ClientsTable'
import Store from '../../components/Store'
import MobileInfo from '../../components/MobileInfo'
import Spinner from '../../components/Spinner'

export default class Clients extends React.Component {
  componentWillMount() {
    this.fetchDataIfNeeded()
  }

  componentDidUpdate(prevProps) {
    this.updateDataOnRouteChange(prevProps)
  }

  updateDataOnRouteChange = (prevProps) => {
    const prevClientId = prevProps.match.params.clientId
    const { clientId } = this.props.match.params
    if (clientId !== prevClientId) this.fetchDataIfNeeded()
  }

  fetchDataIfNeeded = () => {
    if (this.context.isLoading) return

    const { clients, sales, fetchClients, fetchSales } = this.context
    const { clientId } = this.props.match.params

    const shouldFetchClients = !clients.length
    const shouldFetchSales = clientId && !sales.length

    if (shouldFetchClients) fetchClients()
    if (shouldFetchSales) fetchSales()
  }

  render() {
    const { sales, clients } = this.context
    const { match: { params: { clientId } } } = this.props
    const selectedId = Number(clientId)

    const selectedClient= Boolean(clientId) &&
      clients.find(({ id }) => id === selectedId)

    const filteredSales = selectedClient &&
      sales.filter(sale => sale.clientId === selectedId)

    return (
      <div>
        <MobileInfo />

        <h2 className="page-heading">Clients</h2>
        <ClientsTable
          clients={this.context.clients || []}
          selectedId={selectedId}
        />

        <div className="sales">
          {!selectedClient && <h3 className="sales-summary">Select a client above to view their purchases.</h3>}
          {selectedClient && <SalesHeading client={selectedClient} sales={filteredSales}/>}
          {selectedClient && <SalesCards sales={filteredSales} />}
        </div>
      </div>
    )
  }
}

// TODO move these into separate component files
const SalesHeading = ({
  client: { firstName, lastName, company },
  sales
}) => (
  <h3 className="sales-summary">
    <strong>{firstName} {lastName}</strong> of <strong>{company}</strong> has purchased <strong>{sales.length}</strong> widgets
  </h3>
)

const SalesCards = ({ sales = [] }) => (
  sales.length ?
    <ul className="sale-cards">
      {sales.map(
        sale => (
          <li className="sale-card" key={sale.id}>
            <strong>Sale ID: {sale.id}</strong><br />
            {sale.productName} ({sale.size})
          </li>
        )
      )}
    </ul>
    : <Spinner />
)

Clients.contextTypes = Store.childContextTypes
