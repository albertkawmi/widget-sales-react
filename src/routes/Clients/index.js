import Clients from './Clients.route'
export default Clients
