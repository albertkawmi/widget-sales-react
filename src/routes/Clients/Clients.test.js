/* eslint-env jest */
import React from 'react'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Clients from './Clients.route'
import Store from '../../components/Store'

Enzyme.configure({ adapter: new Adapter() })

const store = Enzyme.shallow(<Store />)
const defaultContext = store.instance().getChildContext()

const mockStoreContext = (options = {}) => ({
  ...defaultContext,
  fetchClients: jest.fn(),
  fetchSales: jest.fn(),
  fetchFault: jest.fn(),
  dismissError: jest.fn(),
  ...options
})

describe('<Clients />', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })
  it('calls fetchClients if no existing clients data', () => {
    const context = mockStoreContext()

    expect(
      Enzyme.shallow(
        <Clients match={{ params: {} }} />,
        { context }
      )
    ).toMatchSnapshot()

    expect(context.fetchClients).toHaveBeenCalled()
  })

  it('does not call fetchClients if there is any existing clients data', () => {
    const context = mockStoreContext({ clients: [{ id: 1 }] })

    expect(
      Enzyme.shallow(
        <Clients match={{ params: {} }} />,
        { context }
      )
    ).toMatchSnapshot()

    expect(context.fetchClients).not.toHaveBeenCalled()
  })

  it('calls fetchSales if no existing sales data and there is a :clientId route param', () => {
    const context = mockStoreContext({ clients: [{ id: 1 }] })

    expect(
      Enzyme.shallow(
        <Clients match={{ params: { clientId: 1 } }} />,
        { context }
      )
    ).toMatchSnapshot()

    expect(context.fetchSales).toHaveBeenCalled()
  })

  it('does not call fetchSales if there is any existing sales data', () => {
    const context = mockStoreContext({
      clients: [{ id: 1 }],
      sales: [{ id: 42 }]
    })

    expect(
      Enzyme.shallow(
        <Clients match={{ params: { clientId: 1 } }} />,
        { context }
      )
    ).toMatchSnapshot()

    expect(context.fetchSales).not.toHaveBeenCalled()
  })

  it('calls fetchDataIfNeeded on componentWillMount', () => {
    const context = mockStoreContext({ clients: [{ id: 1 }] })

    const instance = Enzyme.shallow(
      <Clients match={{ params: { clientId: 1 } }} />,
      { context }
    ).instance()

    const spy = jest.spyOn(instance, 'fetchDataIfNeeded')
    instance.componentWillMount({ match: { params: { clientId: 2 } } })
    expect(spy).toHaveBeenCalled()
  })

  it('calls fetchDataIfNeeded on componentDidUpdate if clientId changed', () => {
    const context = mockStoreContext({ clients: [{ id: 1 }] })

    const instance = Enzyme.shallow(
      <Clients match={{ params: { clientId: 1 } }} />,
      { context }
    ).instance()

    const spy = jest.spyOn(instance, 'fetchDataIfNeeded')
    instance.componentDidUpdate({ match: { params: { clientId: 2 } } })
    expect(spy).toHaveBeenCalled()
  })

  it('does not fetchDataIfNeeded on componentDidUpdate if clientId has not changed', () => {
    const context = mockStoreContext({ clients: [{ id: 1 }] })

    const instance = Enzyme.shallow(
      <Clients match={{ params: { clientId: 1 } }} />,
      { context }
    ).instance()

    const spy = jest.spyOn(instance, 'fetchDataIfNeeded')
    instance.componentDidUpdate({ match: { params: { clientId: 1 } } })
    expect(spy).not.toHaveBeenCalled()
  })
})
