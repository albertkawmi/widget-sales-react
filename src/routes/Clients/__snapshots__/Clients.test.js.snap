// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Clients /> calls fetchClients if no existing clients data 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Clients
    match={
        Object {
            "params": Object {},
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <MobileInfo />,
        <h2
          className="page-heading"
>
          Clients
</h2>,
        <ClientsTable
          clients={Array []}
          selectedId={NaN}
/>,
        <div
          className="sales"
>
          <h3
                    className="sales-summary"
          >
                    Select a client above to view their purchases.
          </h3>
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Clients",
          "className": "page-heading",
        },
        "ref": null,
        "rendered": "Clients",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "clients": Array [],
          "selectedId": NaN,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3
              className="sales-summary"
>
              Select a client above to view their purchases.
</h3>,
            false,
            false,
          ],
          "className": "sales",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Select a client above to view their purchases.",
              "className": "sales-summary",
            },
            "ref": null,
            "rendered": "Select a client above to view their purchases.",
            "type": "h3",
          },
          false,
          false,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <MobileInfo />,
          <h2
            className="page-heading"
>
            Clients
</h2>,
          <ClientsTable
            clients={Array []}
            selectedId={NaN}
/>,
          <div
            className="sales"
>
            <h3
                        className="sales-summary"
            >
                        Select a client above to view their purchases.
            </h3>
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Clients",
            "className": "page-heading",
          },
          "ref": null,
          "rendered": "Clients",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "clients": Array [],
            "selectedId": NaN,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3
                className="sales-summary"
>
                Select a client above to view their purchases.
</h3>,
              false,
              false,
            ],
            "className": "sales",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Select a client above to view their purchases.",
                "className": "sales-summary",
              },
              "ref": null,
              "rendered": "Select a client above to view their purchases.",
              "type": "h3",
            },
            false,
            false,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "clients": Array [],
      "dismissError": [Function],
      "error": "",
      "fetchClients": [Function],
      "fetchFault": [Function],
      "fetchSales": [Function],
      "isLoading": false,
      "sales": Array [],
    },
  },
}
`;

exports[`<Clients /> calls fetchSales if no existing sales data and there is a :clientId route param 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Clients
    match={
        Object {
            "params": Object {
              "clientId": 1,
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <MobileInfo />,
        <h2
          className="page-heading"
>
          Clients
</h2>,
        <ClientsTable
          clients={
                    Array [
                              Object {
                                "id": 1,
                              },
                            ]
          }
          selectedId={1}
/>,
        <div
          className="sales"
>
          <SalesHeading
                    client={
                              Object {
                                        "id": 1,
                                      }
                    }
                    sales={Array []}
          />
          <SalesCards
                    sales={Array []}
          />
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Clients",
          "className": "page-heading",
        },
        "ref": null,
        "rendered": "Clients",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "clients": Array [
            Object {
              "id": 1,
            },
          ],
          "selectedId": 1,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            false,
            <SalesHeading
              client={
                            Object {
                                          "id": 1,
                                        }
              }
              sales={Array []}
/>,
            <SalesCards
              sales={Array []}
/>,
          ],
          "className": "sales",
        },
        "ref": null,
        "rendered": Array [
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "client": Object {
                "id": 1,
              },
              "sales": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "sales": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <MobileInfo />,
          <h2
            className="page-heading"
>
            Clients
</h2>,
          <ClientsTable
            clients={
                        Array [
                                    Object {
                                      "id": 1,
                                    },
                                  ]
            }
            selectedId={1}
/>,
          <div
            className="sales"
>
            <SalesHeading
                        client={
                                    Object {
                                                "id": 1,
                                              }
                        }
                        sales={Array []}
            />
            <SalesCards
                        sales={Array []}
            />
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Clients",
            "className": "page-heading",
          },
          "ref": null,
          "rendered": "Clients",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "clients": Array [
              Object {
                "id": 1,
              },
            ],
            "selectedId": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              false,
              <SalesHeading
                client={
                                Object {
                                                "id": 1,
                                              }
                }
                sales={Array []}
/>,
              <SalesCards
                sales={Array []}
/>,
            ],
            "className": "sales",
          },
          "ref": null,
          "rendered": Array [
            false,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "client": Object {
                  "id": 1,
                },
                "sales": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "sales": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "clients": Array [
        Object {
          "id": 1,
        },
      ],
      "dismissError": [Function],
      "error": "",
      "fetchClients": [Function],
      "fetchFault": [Function],
      "fetchSales": [Function],
      "isLoading": false,
      "sales": Array [],
    },
  },
}
`;

exports[`<Clients /> does not call fetchClients if there is any existing clients data 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Clients
    match={
        Object {
            "params": Object {},
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <MobileInfo />,
        <h2
          className="page-heading"
>
          Clients
</h2>,
        <ClientsTable
          clients={
                    Array [
                              Object {
                                "id": 1,
                              },
                            ]
          }
          selectedId={NaN}
/>,
        <div
          className="sales"
>
          <h3
                    className="sales-summary"
          >
                    Select a client above to view their purchases.
          </h3>
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Clients",
          "className": "page-heading",
        },
        "ref": null,
        "rendered": "Clients",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "clients": Array [
            Object {
              "id": 1,
            },
          ],
          "selectedId": NaN,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3
              className="sales-summary"
>
              Select a client above to view their purchases.
</h3>,
            false,
            false,
          ],
          "className": "sales",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Select a client above to view their purchases.",
              "className": "sales-summary",
            },
            "ref": null,
            "rendered": "Select a client above to view their purchases.",
            "type": "h3",
          },
          false,
          false,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <MobileInfo />,
          <h2
            className="page-heading"
>
            Clients
</h2>,
          <ClientsTable
            clients={
                        Array [
                                    Object {
                                      "id": 1,
                                    },
                                  ]
            }
            selectedId={NaN}
/>,
          <div
            className="sales"
>
            <h3
                        className="sales-summary"
            >
                        Select a client above to view their purchases.
            </h3>
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Clients",
            "className": "page-heading",
          },
          "ref": null,
          "rendered": "Clients",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "clients": Array [
              Object {
                "id": 1,
              },
            ],
            "selectedId": NaN,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3
                className="sales-summary"
>
                Select a client above to view their purchases.
</h3>,
              false,
              false,
            ],
            "className": "sales",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Select a client above to view their purchases.",
                "className": "sales-summary",
              },
              "ref": null,
              "rendered": "Select a client above to view their purchases.",
              "type": "h3",
            },
            false,
            false,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "clients": Array [
        Object {
          "id": 1,
        },
      ],
      "dismissError": [Function],
      "error": "",
      "fetchClients": [Function],
      "fetchFault": [Function],
      "fetchSales": [Function],
      "isLoading": false,
      "sales": Array [],
    },
  },
}
`;

exports[`<Clients /> does not call fetchSales if there is any existing sales data 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Clients
    match={
        Object {
            "params": Object {
              "clientId": 1,
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <MobileInfo />,
        <h2
          className="page-heading"
>
          Clients
</h2>,
        <ClientsTable
          clients={
                    Array [
                              Object {
                                "id": 1,
                              },
                            ]
          }
          selectedId={1}
/>,
        <div
          className="sales"
>
          <SalesHeading
                    client={
                              Object {
                                        "id": 1,
                                      }
                    }
                    sales={Array []}
          />
          <SalesCards
                    sales={Array []}
          />
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Clients",
          "className": "page-heading",
        },
        "ref": null,
        "rendered": "Clients",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "clients": Array [
            Object {
              "id": 1,
            },
          ],
          "selectedId": 1,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            false,
            <SalesHeading
              client={
                            Object {
                                          "id": 1,
                                        }
              }
              sales={Array []}
/>,
            <SalesCards
              sales={Array []}
/>,
          ],
          "className": "sales",
        },
        "ref": null,
        "rendered": Array [
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "client": Object {
                "id": 1,
              },
              "sales": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "sales": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <MobileInfo />,
          <h2
            className="page-heading"
>
            Clients
</h2>,
          <ClientsTable
            clients={
                        Array [
                                    Object {
                                      "id": 1,
                                    },
                                  ]
            }
            selectedId={1}
/>,
          <div
            className="sales"
>
            <SalesHeading
                        client={
                                    Object {
                                                "id": 1,
                                              }
                        }
                        sales={Array []}
            />
            <SalesCards
                        sales={Array []}
            />
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Clients",
            "className": "page-heading",
          },
          "ref": null,
          "rendered": "Clients",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "clients": Array [
              Object {
                "id": 1,
              },
            ],
            "selectedId": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              false,
              <SalesHeading
                client={
                                Object {
                                                "id": 1,
                                              }
                }
                sales={Array []}
/>,
              <SalesCards
                sales={Array []}
/>,
            ],
            "className": "sales",
          },
          "ref": null,
          "rendered": Array [
            false,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "client": Object {
                  "id": 1,
                },
                "sales": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "sales": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "clients": Array [
        Object {
          "id": 1,
        },
      ],
      "dismissError": [Function],
      "error": "",
      "fetchClients": [Function],
      "fetchFault": [Function],
      "fetchSales": [Function],
      "isLoading": false,
      "sales": Array [
        Object {
          "id": 42,
        },
      ],
    },
  },
}
`;
